replica/database.cc

```diff
@@ -2849,14 +2852,29 @@ flat_mutation_reader_v2 make_multishard_streaming_reader(distributed<replica::da
             streamed_mutation::forwarding,
             mutation_reader::forwarding fwd_mr) {
         auto table_id = s->id();
-        return make_multishard_combining_reader_v2(make_shared<streaming_reader_lifecycle_policy>(db, table_id), std::move(s), std::move(permit), pr, ps,
+        return make_multishard_combining_reader_v2(make_shared<replica::streaming_reader_lifecycle_policy>(db, table_id), std::move(s), std::move(permit), pr, ps,
                 std::move(trace_state), fwd_mr);
     });
     auto&& full_slice = schema->full_slice();
     return make_flat_multi_range_reader(schema, std::move(permit), std::move(ms),
             std::move(range_generator), std::move(full_slice), {}, mutation_reader::forwarding::no);
 }
 
+flat_mutation_reader_v2 make_multishard_streaming_reader(distributed<replica::database>& db,
+        schema_ptr schema, reader_permit permit,
+        const dht::partition_range& range, noncopyable_function<bool(const dht::decorated_key&)>&& filter)
+{
+    const auto& table_id = schema->id();
+    const auto& full_slice = schema->full_slice();
+    auto multishard_reader = make_multishard_combining_reader_v2(
+        make_shared<replica::streaming_reader_lifecycle_policy>(db, table_id),
+        std::move(schema),
+        std::move(permit),
+        range,
+        full_slice);
+    return make_filtering_reader(std::move(multishard_reader), std::move(filter));
+}
```

Comment:
I think it would be better if you selected the parameters and added the filtering on top in the caller.

Reply:
Or in other words, not add a new overload but just pick the params for the existing `make_...` method.

Reply:
You mean a `range_generator` returning one single range, which is the entire range from the repair master? It feels a bit awkward to me, and it leaves some overhead from `make_flat_multi_range_reader`. An overload with explicit `dht::partition_range` parameter is more straightforward. I can move filtering to the caller, though.

Reply:
Ah, I missed that we have a range generator parameter. Indeed that merits another overload, with the filtering stage left for the caller to do.